### PR TITLE
[gql][feature] suinsRegistration connection

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Looks like the NameServiceName node is not fully defined in the draft schema. Do we want this to follow the standard we've set for other "special" move objects like Coin, CoinMetadata, and StakedSui, or are we fine with having it represented as a string?

If the former, the query syntax would be something like
```
{
  address(
    address: "0x"
  ) {
    nameServiceNames(first: 1) {
      nodes {
        asMoveObject {
          contents {
            json
            type {
              repr
            }
            data
          }
        }
        domain
      }
    }
  }
}
```

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Review by amnn — COMMENTED** [review]
From an implementation perspective this looks great!  In terms of your question, I think this comes down to the preferences of @Jordan-Mysten, @hayes-mysten, or @manolisliolios as potential users of this API, and whether it is useful to get access to the `SuinsRegistration` as a `MoveObject` or whether it's only ever useful to get access to the name directly.

If the latter, then we should just have this API return the name as a string, if the former, then I like the approach you've taken, although we should have the name of the GraphQL type match the name of the Move type.

**Comment by wlmyng** [comment]
From discussions on slack, even though we only find the 'domain' field meaningful right now, having the return type as a special move object instead of a string will make extending this field easier later on, say if we want to link to or show the image, other ipfs addresses, etc.

There's a separate concern about baking a Mysten product into the graphql rpc, but this can be addressed pretty easily by pulling the implementation out into a suins-specific subgraph

File: crates/sui-graphql-rpc/schema/current_progress_schema.graphql [hunk 1]
```diff
@@ -646,15 +646,15 @@ type Event {
 	data: MoveData!
 	"""
 	Representation of a Move value in JSON, where:
-	
+
```
Comment (amnn) [inline crates/sui-graphql-rpc/schema/current_progress_schema.graphql:649]: nit: I think you probably want to undo these changes?  Unfortunately, the schema generator leaves trailing whitespace, and your editor is probably cleaning it up when you open this file, but these will just get added back or cause a merge conflict if you check in the change.

File: crates/sui-graphql-rpc/schema/draft_target_schema.graphql [hunk 2]
```diff
@@ -1522,14 +1522,14 @@ type MoveModuleEdge {
...
+type SuinsRegistrationConnection {
+  edges: [SuinsRegistrationEdge!]!
+  nodes: [SuinsRegistration!]!
   pageInfo: PageInfo!
 }
 
-type NameServiceNameEdge {
+type SuinsRegistrationEdge {
   cursor: String
   node: String!  # is this correct?
```
Comment (amnn) [inline crates/sui-graphql-rpc/schema/draft_target_schema.graphql:1534]: This should now be a `SuinsRegistration` as well, right?

File: crates/sui-graphql-rpc/schema/draft_target_schema.graphql [hunk 3]
```diff
@@ -476,13 +476,13 @@ interface IOwner {
     filter: DynamicFieldFilter,
   ): DynamicFieldConnection!
 
-  defaultNameServiceName: String
-  nameServiceNames(
+  defaultSuinsRegistration: String
+  SuinsRegistrations(
```
Comment (amnn) [inline crates/sui-graphql-rpc/schema/draft_target_schema.graphql:480]: nit, on capitalisation here:

```suggestion
  defaultSuinsRegistration: String
  suinsRegistrations(
```

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 4]
```diff
@@ -1485,6 +1486,73 @@ impl PgManager {
...
+            let object = Object::try_from(stored_obj)?;
+
+            let move_object = MoveObject::try_from(&object).map_err(|_| {
+                Error::Internal(format!(
+                    "Expected {} to be a suinsRegistration object, but it's not an object",
+                    object.address,
+                ))
+            })?;
+
+            let nsn = SuinsRegistration::try_from((&move_object, &struct_tag)).map_err(|_| {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:1539]: nit: `nsn` probably doesn't make sense as a name for this anymore!

File: crates/sui-graphql-rpc/src/functional_group.rs [hunk 5]
```diff
@@ -94,14 +94,14 @@ fn functional_groups() -> &'static BTreeMap<(&'static str, &'static str), Functi
             (("Object", "balance"), G::Coins),
             (("Object", "balanceConnection"), G::Coins),
             (("Object", "coinConnection"), G::Coins),
-            (("Object", "defaultNameServiceName"), G::NameService),
+            (("Object", "defaultSuinsRegistration"), G::NameService),
             (("Object", "dynamicField"), G::DynamicFields),
             (("Object", "dynamicFieldConnection"), G::DynamicFields),
             // (("Object", "nameServiceConnection"), G::NameService),
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/functional_group.rs:100]: Looks like you would also need to uncomment this and rename it to `suinsRegistrations`?

File: crates/sui-graphql-rpc/src/types/address.rs [hunk 6]
```diff
@@ -146,16 +147,26 @@ impl Address {
...
-    // pub async fn name_service_connection(
-    //     &self,
-    //     first: Option<u64>,
-    //     after: Option<String>,
-    //     last: Option<u64>,
-    //     before: Option<String>,
-    // ) -> Result<Option<Connection<String, NameService>>> {
-    //     unimplemented!()
-    // }
+    pub async fn name_service_names(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/address.rs:150]: Should this be `suinsRegistrations`?

File: crates/sui-graphql-rpc/src/types/name_service_name.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/name_service_name.rs:1]: nit: name change to match type it defines -- `suins_registration.rs`.

File: crates/sui-graphql-rpc/src/types/name_service_name.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+use async_graphql::*;
+use move_core_types::language_storage::StructTag;
+use serde::{Deserialize, Serialize};
+use sui_json_rpc::name_service::Domain;
+
+use super::move_object::MoveObject;
+
+#[derive(Clone, Serialize, Deserialize)]
+pub(crate) struct NativeSuinsRegistration {
+    pub id: sui_types::id::UID,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/name_service_name.rs:13]: nit: import?

File: crates/sui-graphql-rpc/src/types/name_service_name.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    pub domain: Domain,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/name_service_name.rs:14]: Do we have to take the dependency on `sui_json_rpc` here?

File: crates/sui-graphql-rpc/src/types/name_service_name.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        &self.native.domain_name
+    }
+
+    /// Convert the SuinsRegistration object into a Move object
+    async fn as_move_object(&self) -> &MoveObject {
+        &self.super_
+    }
+}
+
+impl TryFrom<(&MoveObject, &StructTag)> for SuinsRegistration {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/name_service_name.rs:47]: Don't implement conversion traits against tuples -- that's just a regular function that you've made difficult to call!

Comment (amnn) [reply crates/sui-graphql-rpc/src/types/name_service_name.rs:47]: But at the same time, I don't fully understand why you need to pass the `StructTag` here?  Surely that would be a constant that represents the `SuinsRegistration` type?

**Review by amnn — COMMENTED** [review]
I think this is mostly here, but some nits.  Main thing blocking the accept is the logic around conversion from `MoveObject` to `SuinsRegistration` -- I don't think I fully understand what's going on with that `StructTag` parameter but it seems fishy to me, could you take another look?

File: crates/sui-graphql-rpc/src/types/name_service_name.rs (hunk 9 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/name_service_name.rs:47]: @manolisliolios iiuc the mainnet package address for suins is fixed, but testnet is not and it changes after every testnet wipe?

Comment (amnn) [reply crates/sui-graphql-rpc/src/types/name_service_name.rs:47]: Ah got it -- that is annoying, but okay, then this shouldn't be a `TryFrom` impl, it should just be a function:

```
impl SuinsRegistration {
    fn try_from(object: &MoveObject, tag: &StructTag) -> Result<Self, Error> { /* ... */ }
}
```

**Review by amnn — APPROVED** [review]
Let me also accept to unblock this change -- just make sure to replace the `TryFrom<(...)>`!

**Comment by wlmyng** [comment]
I'll address the Owner.defaultNameServiceName and Query.resolveNameServiceAddress renaming in a subsequent PR. I'm hesitant to apply the same naming scheme since the Owner.defaultNameServiceName refers to the reverse registry to look up the default domain set for some address. Similarly, Query.resolveNameServiceAddress checks the registry for the target address of the domain. Since the underlying object(s) are not the SuinsRegistrationNFT but rather the respective registries, I think it'd be confusing to use similar language.

Thinking Query.resolveNameServiceAddress -> Query.resolveSuinsDomainTargetAddress and
Owner.defaultNameServiceName -> Owner.defaultSuinsDomain

(since the underlying "object" is a Domain dynamic field on the registry)
